Mocked BLE-WiFi commissioning test

arkq opened the pull request:
#### Summary

Add support for mocked BLE-WiFi commissioning on Linux when running test suite.

#### Testing

Tested locally with:

```sh
scripts/tests/run_test_suite.py --target TestBasicInformation run --ble-wifi
```

CI will verify that as well.

> [!NOTE]
> The `Test Suites - Linux (no-ble-no-shell-tsan-clang)` CI check waits for report because there is no  `no-ble-no-shell-tsan-clang` matrix any more. This action is now named (due to matrix change) as `Test Suites - Linux (no-shell-tsan-clang)`. Someone will have to updated required checks in project's settings after this PR will land into master.

#### Readability checklist

The checklist below will help the reviewer finish PR review in time and keep the
code readable:

-   [x] PR title is
        [descriptive](https://project-chip.github.io/connectedhomeip-doc/pull_request_guidelines.html#title-formatting)
-   [x] Apply the
        [_“When in Rome…”_](https://project-chip.github.io/connectedhomeip-doc/style/CODING_STYLE_GUIDE.html)
        rule (coding style)
-   [x] PR size is short
-   [x] Try to avoid "squashing" and "force-update" in commit history
-   [ ] CI time didn't increase

See: [PULL_REQUEST_GUIDELINES.md](../docs/pull_request_guidelines.md)

jlatusek committed:
Add qemu testing to CI

arkq committed:
Add help message for --ble-adapter chip-tool option

arkq committed:
Use BlueZ and WpaSupplicant D-Bus mocks for testing

arkq committed:
Add BlueZoo to scripts Python requirements

arkq committed:
Add BLE-WiFi commissioning test to CI

restyled-commits committed:
Fix comment issues

arkq commented on scripts/tests/chiptest/linux.py:
```diff
@@ -169,6 +174,157 @@ def ShutdownNamespaceForTestExecution():
+                         stderr=subprocess.DEVNULL)
+        os.environ["DBUS_SYSTEM_BUS_ADDRESS"] = self.ADDRESS
+        # Wait for the bus to start.
+        time.sleep(0.5)
```
Changed to wait for a message from `dbus-daemon`.

arkq commented on scripts/tests/run_test_suite.py:
```diff
@@ -367,9 +373,18 @@
+            bluetooth = chiptest.linux.BluetoothMock()
+            wifi = chiptest.linux.WpaSupplicantMock()
+            time.sleep(0.5)  # Give the mock servers time to start
+            ble_adapter_app = 0   # Bind app to the first BLE adapter
+            ble_adapter_tool = 1  # Bind tool to the second BLE adapter
```
Test setup requires at least 2 controllers. This requirement will not change, so it's not required to make selection dynamic.

arkq committed:
Run Linux-specific code only on Linux

arkq committed:
Merge branch 'master' into ble-commussioning-ci

arkq committed:
Disable TSAN for BLE-WiFi CI test

arkq committed:
Merge branch 'master'

bzbarsky-apple commented on scripts/tests/chiptest/test_definition.py:
```diff
@@ -360,23 +364,28 @@ def Run(self, runner, apps_register, paths: ApplicationPaths, pics_file: str,
-
-                    app = App(runner, path)
+                    else:
+                        key = os.path.basename(command[-1])
```
Why are we ignoring the `key` from our iteration, then?

bzbarsky-apple commented on scripts/tests/chiptest/test_definition.py:
```diff
@@ -408,21 +417,34 @@ def Run(self, runner, apps_register, paths: ApplicationPaths, pics_file: str,
-                pairing_cmd = paths.chip_tool_with_python_cmd + ['pairing', 'code', TEST_NODE_ID, setupCode]
+                pairing_server_args = []
+                if ble_controller_tool is not None:
+                    # Explicitly use ble-wifi pairing to force the use of BLE
+                    # exchange and AP association. The code-wifi pairing skips
+                    # BLE commissioning if devices can be commissioned over IP
+                    # network.
```
Can we fix the server to not be advertising on the IP network?   It's _really_ weird (like: forbidden by the spec) to be advertising both on the IP network and over BLE.

arkq replied:
I've changed the test setup. Now, the IP on the veth link (which name is prefixed with `wl` so it is recognized as wlan) is assigned only after successful wifi acquisition (done by the mocked WPA supplicant). With that setup our Matter application is not linked with controller over veth, so the code-wifi pairing successfully selects BLE for commissioning.

bzbarsky-apple commented on scripts/tests/run_test_suite.py:
```diff
@@ -295,10 +295,16 @@ def cmd_list(context):
+    is_flag=True,
+    default=False,
+    show_default=True,
+    help='Use a Bluetooth and WiFi mock servers to perform BLE-WiFi commissioning')
```
```suggestion
    help='Use Bluetooth and WiFi mock servers to perform BLE-WiFi commissioning')
```

bzbarsky-apple commented on scripts/tests/run_test_suite.py:
```diff
@@ -295,10 +295,16 @@ def cmd_list(context):
     default=0,
     show_default=True,
     help='Number of tests that are expected to fail in each iteration.  Overall test will pass if the number of failures matches this.  Nonzero values require --keep-going')
+@click.option(
```
Should this be conditioned on sys.platform?

arkq replied:
Added a help message which indicates that this option is available only on Linux, and a runtime check to fail-fast if someone passes that option on other platforms.

arkq committed:
Add review suggestions

arkq committed:
Merge branch 'master'

arkq committed:
Fix WPA supplicant AP association mock logic

arkq committed:
[Linux] Improve isolated test network management

arkq committed:
Merge branch 'master' into ble-commussioning-ci

restyled-commits committed:
Restyled by isort

arkq committed:
Merge branch 'master' into ble-commussioning-ci

arkq committed:
Fix workflow syntax issue

enkiusz commented on src/platform/Linux/bluez/BluezObjectManager.cpp:
```diff
@@ -274,9 +274,9 @@ CHIP_ERROR BluezObjectManager::SetupObjectManager()
-        mConnection.get(), G_DBUS_OBJECT_MANAGER_CLIENT_FLAGS_NONE, BLUEZ_INTERFACE, "/",
-        bluez_object_manager_client_get_proxy_type, nullptr /* unused user data in the proxy type func */,
-        nullptr /* destroy notify */, nullptr /* cancellable */, &err.GetReceiver()));
+        mConnection.get(), G_DBUS_OBJECT_MANAGER_CLIENT_FLAGS_NONE, "org.bluez", "/", bluez_object_manager_client_get_proxy_type,
```
Why do we hardcode this ?

arkq commented on src/platform/Linux/bluez/BluezObjectManager.cpp:
```diff
@@ -274,9 +274,9 @@ CHIP_ERROR BluezObjectManager::SetupObjectManager()
-        mConnection.get(), G_DBUS_OBJECT_MANAGER_CLIENT_FLAGS_NONE, BLUEZ_INTERFACE, "/",
-        bluez_object_manager_client_get_proxy_type, nullptr /* unused user data in the proxy type func */,
-        nullptr /* destroy notify */, nullptr /* cancellable */, &err.GetReceiver()));
+        mConnection.get(), G_DBUS_OBJECT_MANAGER_CLIENT_FLAGS_NONE, "org.bluez", "/", bluez_object_manager_client_get_proxy_type,
```
I can add this to some defines. My point here was to not use interface name for service name (the same value, but semantically incorrect)

enkiusz replied:
We could add it to `src/platform/Linux/bluez/Types.h` alongside `BLUEZ_INTERFACE`

arkq committed:
Apply comment suggestions

bzbarsky-apple commented on scripts/tests/chiptest/__init__.py:
```diff
@@ -18,13 +18,26 @@
 
+__all__ = [
+    "TestTarget",
+    "TestDefinition",
```
What happened to AllTests?

arkq replied:
I've removed it since there is no such class in that file (error reported by pylance).

arkq committed:
Merge branch 'master' into ble-commussioning-ci

arkq committed:
Merge branch 'master'

arkq committed:
[Linux] Fix SIGSEGV when returning error from call handler

The call handler needs to return TRUE (G_DBUS_METHOD_INVOCATION_HANDLED)
if the call was handed. Returning FALSE means that the call will be
handled internally by glib. In our scenario returning FALSE results in
handling call twice which leads to double-free.

arkq committed:
Merge branch 'master'

arkq committed:
Wait for the BLE adapters to be ready

arkq committed:
Enable verbose output for CI debugging

arkq committed:
Enforce bluezoo >= 1.0.1

arkq committed:
Update constraints

arkq committed:
Remove debug log level from test command

bzbarsky-apple reviewed (approved):
The non-Linux-specific bits look good.  I can't comment on the Linux bits.

andy31415 committed:
Merge branch 'master' into ble-commussioning-ci
